XenAdmin/SettingsPanels/VBDEditPage.cs

```diff
@@ -149,16 +150,30 @@ public void UpdateDevicePositions(Session session)
                 {
                     devicePositionComboBox.Items.Clear();
 
-                    devicePositionComboBox.Items.AddRange(devices.ToArray());
+                    var comboBoxItems = vm.Connection.ResolveAll(vm.VBDs).Select(d => new DevicePositionComboBoxItem(d.userdevice, d));
+
+                    foreach (var devicePosition in devices)
+                    {
+                        var comboBoxItemWithVdi = comboBoxItems.FirstOrDefault(ci => ci.VBD.userdevice == devicePosition);
+                        if (comboBoxItemWithVdi != null)
+                        {
+                            devicePositionComboBox.Items.Add(comboBoxItemWithVdi);
+                        }
+                        else
+                        {
+                            devicePositionComboBox.Items.Add(new DevicePositionComboBoxItem(devicePosition));
+                        }
+                    }
 
                     // Make sure the userdevice for the selected VBD is the one selected in the combobox
-                    foreach (String position in devicePositionComboBox.Items)
+                    foreach (DevicePositionComboBoxItem item in devicePositionComboBox.Items)
                     {
-                        if (position != vbd.userdevice)
-                            continue;
+                        if (item.VBD != null && item.VBD.userdevice == vbd.userdevice)
+                        {
+                            devicePositionComboBox.SelectedItem = item;

```

Comment:
We may want to select the item in the loop above that populates the combobox to avoid the re-iteration.